Modified the my_none? and my_all? methods to work in all familiar test cases. !!Read comment!!

yirano opened the pull request:
first pull request.

mcrd25 commented on enumerable.rb:
```diff
@@ -114,9 +114,23 @@ def my_map
   end
 
   def my_map_proc
```
This method is not necessary. The handling of proc should be don in the `my_map` method. As it stands your my_map handles procs well.

yirano replied:
Deleted this as well! Thank you!

mcrd25 commented on enumerable.rb:
```diff
@@ -114,9 +114,23 @@ def my_map
+    return arr
   end
 
   def my_map_proc_bloc
```
This method is not necessary. The handling of proc should be don in the `my_map` method. As it stands your my_map handles procs and blocks as should.

yirano replied:
I removed this method. Thank you!

mcrd25 commented on enumerable.rb:
```diff
@@ -114,9 +114,23 @@ def my_map
+    end
   end
 
   def my_inject initial_value = nil, symbol = nil
```
Your inject does not currently work as should. See main comment for more detail

mcrd25 replied:
Please consider that if there is no initial value in call or if only symbol argument is passed then that means the accumulator starts at the first value in passed array otherwise it sets initial accumulator value as `initial_value`.

Also if initial accumulator sets to first value of array then you iterate from second value [1..-1] in array otherwise if argument for initial value was passed you iterate from first.

mcrd25 replied:
You need to recheck how you do handle these conditions. you seem to have the main idea though!

yirano replied:
The inject should work the way it was intended! I think I might be missing a portion of it but I'd like to make sure I'm on the right track before I proceed.

yirano commented:
@mcrd25 This has got to be **THE BEST** code review I have ever received. Thank you so, so much for taking the time for a thorough explanation of what I should fix. I was a little anxious about submitting this because I wasn't even sure if I was doing any of this right.

Thank you again!! SO MUCH! 

I'll get to fixing these changes right away!

yirano commented:
- modified a few methods to return an enumerator if it has no block. 
- I think I corrected the inject method. I'm pretty sure one other thing needs to be implemented when an initial value is given. I just want to make sure I'm on the right track before I commit to proceed any further.

samgaco commented on enumerable.rb:
```diff
@@ -89,97 +159,108 @@ def my_none?
     end
-  end
+    memo
 end
```
### 1.  there is an `end` lacking here, to leave the module enclosed

samgaco commented on enumerable.rb:
```diff
@@ -89,97 +159,108 @@ def my_none?
+  def my_inject(memo = nil)
+    if memo.includes?(Symbol) == false
+      self.each do |x| 
+        if memo = memo.nil?
```
#### 2.  You should use `== `  for if conditions  ` memo == memo.nil? `

samgaco commented on enumerable.rb:
```diff
@@ -89,97 +159,108 @@ def my_none?
+        end
+    else
+      p "HI"
     end
```
#### 3. The `else` part of this `if` might be unnecessary

yirano replied:
Oh! Sorry about this. This was meant to be a test case to see what would happen if the statement failed.

samgaco commented on enumerable.rb:
```diff
@@ -89,97 +159,108 @@ def my_none?
-      end
-      initial_value
+  def my_inject(memo = nil)
+    if memo.includes?(Symbol) == false
```
#### 5.  If you want to pass an if condition if the statement is false, ` !memo.includes?(Symbol) ` would look cleaner

samgaco commented on enumerable.rb:
```diff
@@ -2,70 +2,113 @@
   
   # Everything below this comment does not have a complete working code for Hash
-  def my_all?
+  def my_all?(pattern = nil)
```
#### 6. When no block or argument is given, your method should return `true` if none of the elements in the collection is nil or false, in your case, the method is not working when passing an array with integers and strings:


your code tested against the actual one:

```
irb(main):245:0> ["ok",1].my_all?
=> false
irb(main):246:0> ["ok",1].all?
=> true
irb(main):247:0> ["ok",1, false].my_all?
=> false
irb(main):248:0> ["ok",1, false].all?
=> false
```


#### 7. when passing a pattern, it should be able to detect if all the elements of the collection follow such pattern.

```
irb(main):253:0> [":D"].all?(":D")
=> true
irb(main):254:0> [":D"].my_all?(":D")
=> true
irb(main):255:0> [":D",1].my_all?(":D")
=> true
irb(main):256:0> [":D",1].all?(":D")
=> false
```

samgaco commented on enumerable.rb:
```diff
@@ -2,70 +2,113 @@
   end
 
-  def my_any?
+  def my_any?(pattern = nil)
```
#### 8. Same problems here as in my_all?  but applied for my_any? 


```
irb(main):261:0> [1,1,1,1].any?
=> true
irb(main):262:0> [1,1,1,1].my_any?
=> false
```


and


```
irb(main):268:0> ["ok"].any?("ok")
=> true
irb(main):269:0> ["ok"].my_any?("ok")
=> false

```

samgaco commented on enumerable.rb:
```diff
@@ -75,11 +118,38 @@ def my_any?
   end
 
-  def my_none?
+  def my_none?(pattern = nil)
```
#### 9. Should work when passing a class

```
irb(main):270:0> [false, "rr"].none?(Integer)
=> true
irb(main):271:0> [false, "rr"].my_none?(Integer)
=> false
```


and same as in my_any? and my_all? when passing any pattern (not a class or a regular expression):

```
irb(main):272:0> ["o"].none?("o")
=> false
irb(main):273:0> ["o"].my_none?("o")
=> true

```

yirano commented:
The more I dug into my previous code the more stringy and messy it seemed to get.

Therefore, I recreated the entire code structure. There are Stickler errors showing and if it is absolutely necessary that I pass those tests I will correct them.

maelfosso commented on enumerable.rb:
```diff
@@ -1,185 +1,225 @@
-    end
-    self
+    block_given? ? my_each { |x| yield(x) } : enum_for(:my_each_with_index)
   end
```
The result are not the one I expect
```ruby
irb(main):018:0> array
=> [6, 5, 8, 2, 7, 2, 1, 8, 5, 4, 2, 7, 7, 6, 6, 4, 5, 8, 2, 5, 3, 8, 5, 7, 6, 5, 8, 5, 0, 1, 6, 1, 0, 0, 2, 6, 0, 8, 4, 0, 6, 8, 4, 5, 8, 6, 0, 3, 6, 3, 2, 4, 2, 7, 2, 4, 2, 7, 6, 1, 7, 1, 2, 4, 6, 2, 3, 0, 2, 1, 3, 4, 8, 3, 4, 5, 5, 0, 5, 2, 6, 6, 0, 6, 8, 2, 8, 3, 3, 8, 4, 1, 2, 4, 6, 4, 0, 6, 1, 4]
irb(main):019:0> my_each_output = ''
=> ""
irb(main):020:0> block = proc { |num, idx| my_each_output += "Num: #{num}, idx: #{idx}\n" }
=> #<Proc:0x0000558ed1d0f2e0@(irb):20>
irb(main):021:0> array.each_with_index(&block)
=> [6, 5, 8, 2, 7, 2, 1, 8, 5, 4, 2, 7, 7, 6, 6, 4, 5, 8, 2, 5, 3, 8, 5, 7, 6, 5, 8, 5, 0, 1, 6, 1, 0, 0, 2, 6, 0, 8, 4, 0, 6, 8, 4, 5, 8, 6, 0, 3, 6, 3, 2, 4, 2, 7, 2, 4, 2, 7, 6, 1, 7, 1, 2, 4, 6, 2, 3, 0, 2, 1, 3, 4, 8, 3, 4, 5, 5, 0, 5, 2, 6, 6, 0, 6, 8, 2, 8, 3, 3, 8, 4, 1, 2, 4, 6, 4, 0, 6, 1, 4]
irb(main):022:0> each_output = my_each_output.dup
=> "Num: 6, idx: 0\nNum: 5, idx: 1\nNum: 8, idx: 2\nNum: 2, idx: 3\nNum: 7, idx: 4\nNum: 2, idx: 5\nNum: 1, idx: 6\nNum: 8, idx: 7\nNum: 5, idx: 8\nNum: 4, idx: 9\nNum: 2, idx: 10\nNum: 7, idx: 11\nNum: 7, idx: 12\nNum: 6, idx: 13\nNum: 6, idx: 14\nNum: 4, idx: 15\nNum: 5, idx: 16\nNum: 8, idx: 17\nNum: 2, idx: 18\nNum: 5, idx: 19\nNum: 3, idx: 20\nNum: 8, idx: 21\nNum: 5, idx: 22\nNum: 7, idx: 23\nNum: 6, idx: 24\nNum: 5, idx: 25\nNum: 8, idx: 26\nNum: 5, idx: 27\nNum: 0, idx: 28\nNum: 1, idx: 29\nNum: 6, idx: 30\nNum: 1, idx: 31\nNum: 0, idx: 32\nNum: 0, idx: 33\nNum: 2, idx: 34\nNum: 6, idx: 35\nNum: 0, idx: 36\nNum: 8, idx: 37\nNum: 4, idx: 38\nNum: 0, idx: 39\nNum: 6, idx: 40\nNum: 8, idx: 41\nNum: 4, idx: 42\nNum: 5, idx: 43\nNum: 8, idx: 44\nNum: 6, idx: 45\nNum: 0, idx: 46\nNum: 3, idx: 47\nNum: 6, idx: 48\nNum: 3, idx: 49\nNum: 2, idx: 50\nNum: 4, idx: 51\nNum: 2, idx: 52\nNum: 7, idx: 53\nNum: 2, idx: 54\nNum: 4, idx: 55\nNum: 2, idx: 56\nNum: 7, idx: 57\nNum: 6, idx: 58\nNum: 1, idx: 59\nNum: 7, idx: 60\nNum: 1, idx: 61\nNum: 2, idx: 62\nNum: 4, idx: 63\nNum: 6, idx: 64\nNum: 2, idx: 65\nNum: 3, idx: 66\nNum: 0, idx: 67\nNum: 2, idx: 68\nNum: 1, idx: 69\nNum: 3, idx: 70\nNum: 4, idx: 71\nNum: 8, idx: 72\nNum: 3, idx: 73\nNum: 4, idx: 74\nNum: 5, idx: 75\nNum: 5, idx: 76\nNum: 0, idx: 77\nNum: 5, idx: 78\nNum: 2, idx: 79\nNum: 6, idx: 80\nNum: 6, idx: 81\nNum: 0, idx: 82\nNum: 6, idx: 83\nNum: 8, idx: 84\nNum: 2, idx: 85\nNum: 8, idx: 86\nNum: 3, idx: 87\nNum: 3, idx: 88\nNum: 8, idx: 89\nNum: 4, idx: 90\nNum: 1, idx: 91\nNum: 2, idx: 92\nNum: 4, idx: 93\nNum: 6, idx: 94\nNum: 4, idx: 95\nNum: 0, idx: 96\nNum: 6, idx: 97\nNum: 1, idx: 98\nNum: 4, idx: 99\n"
irb(main):023:0> my_each_output = ''
=> ""
irb(main):024:0> array.my_each_with_index(&block)
=> [6, 5, 8, 2, 7, 2, 1, 8, 5, 4, 2, 7, 7, 6, 6, 4, 5, 8, 2, 5, 3, 8, 5, 7, 6, 5, 8, 5, 0, 1, 6, 1, 0, 0, 2, 6, 0, 8, 4, 0, 6, 8, 4, 5, 8, 6, 0, 3, 6, 3, 2, 4, 2, 7, 2, 4, 2, 7, 6, 1, 7, 1, 2, 4, 6, 2, 3, 0, 2, 1, 3, 4, 8, 3, 4, 5, 5, 0, 5, 2, 6, 6, 0, 6, 8, 2, 8, 3, 3, 8, 4, 1, 2, 4, 6, 4, 0, 6, 1, 4]
irb(main):025:0> 
```

yirano replied:
If you look at The Odin Project, procs are not required to work on this particular recreated method. 

The Odin Project **specifically asks** that the my_map method accepts but doesn't require it for others. So I'd like to be opted out of this requirement for that reason.

maelfosso commented on enumerable.rb:
```diff
@@ -1,185 +1,225 @@
     end
-    return run
+    result
   end
```
```ruby
irb(main):028:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):029:0> words.all?(/d/)
=> true
irb(main):030:0> words.my_all?(/d/)
=> false
irb(main):031:0> 
```

maelfosso replied:
You've an error instead of returning `true or false`
```ruby
irb(main):034:0> array
=> [6, 5, 8, 2, 7, 2, 1, 8, 5, 4, 2, 7, 7, 6, 6, 4, 5, 8, 2, 5, 3, 8, 5, 7, 6, 5, 8, 5, 0, 1, 6, 1, 0, 0, 2, 6, 0, 8, 4, 0, 6, 8, 4, 5, 8, 6, 0, 3, 6, 3, 2, 4, 2, 7, 2, 4, 2, 7, 6, 1, 7, 1, 2, 4, 6, 2, 3, 0, 2, 1, 3, 4, 8, 3, 4, 5, 5, 0, 5, 2, 6, 6, 0, 6, 8, 2, 8, 3, 3, 8, 4, 1, 2, 4, 6, 4, 0, 6, 1, 4]
irb(main):035:0> array.all?(3)
=> false
irb(main):036:0> array.my_all?(3)
Traceback (most recent call last):
        7: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        6: from (irb):36
        5: from enumerable.rb:43:in `my_all?'
        4: from enumerable.rb:10:in `my_each'
        3: from enumerable.rb:10:in `each'
        2: from enumerable.rb:11:in `block in my_each'
        1: from enumerable.rb:44:in `block in my_all?'
NoMethodError (undefined method `match?' for 6:Integer)
irb(main):037:0> 
```

maelfosso commented on enumerable.rb:
```diff
@@ -1,185 +1,225 @@
     end
-    return found
+    result
   end
```
```ruby
irb(main):039:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):040:0> words.none?(5)
=> true
irb(main):041:0> words.my_none?(5)
Traceback (most recent call last):
        8: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        7: from (irb):41
        6: from enumerable.rb:94:in `my_none?'
        5: from enumerable.rb:10:in `my_each'
        4: from enumerable.rb:10:in `each'
        3: from enumerable.rb:11:in `block in my_each'
        2: from enumerable.rb:95:in `block in my_none?'
        1: from enumerable.rb:95:in `match?'
TypeError (wrong argument type Integer (expected Regexp))
irb(main):042:0> 
```

maelfosso commented on enumerable.rb:
```diff
@@ -1,185 +1,225 @@
+    elsif symbol.is_a?(Symbol)
+      symbol_detected(memo, symbol, arr)
     end
   end
```
```ruby
irb(main):042:0> search = proc { |memo, word| memo.length > word.length ? memo : word }
=> #<Proc:0x0000558ed1ce9fb8@(irb):42>
irb(main):043:0> words.inject(&search)
=> "blade"
irb(main):044:0> words.my_inject(&search)
Traceback (most recent call last):
        8: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        7: from (irb):44
        6: from enumerable.rb:150:in `my_inject'
        5: from enumerable.rb:10:in `my_each'
        4: from enumerable.rb:10:in `each'
        3: from enumerable.rb:11:in `block in my_each'
        2: from enumerable.rb:150:in `block in my_inject'
        1: from (irb):42:in `block in irb_binding'
NoMethodError (undefined method `length' for 0:Integer)
irb(main):045:0> 
```

yirano replied:
According to the actual project guideline, this recreated method is also not required to accept procs. I could be wrong about this so please correct me if I am.

yirano commented:
Some of the methods, as outlined on this project outline, are not required to accept procs. That was required _**specifically**_ for the my_map method and not the rest. I have therefore left some methods as is.

yirano commented:
> my_each_with_index calls the given block once for each element in self

It's so funny because the more I dig into my code the more I mess it up 🤣. The Regex was working for my_all? and my_none? but I'm seeing now that it doesn't. I'll just redo this project again so I'm going to close this PR and make a new one.

I do have a little concern about the Stickler error though. In some instances the error showed up because it translated some of the "self" used as redundant.... but in actuality it wasn't. It had to be there and actually did something. It wasn't written "self.[something here]" but it was just "size". What do I do about Stickler errors like those? Some of them are more opinions and preferences like "each is preferred over for loops" but it works the same way. Not to mention that in SOME cases, the for loops kind of looks a little better if not more efficient? IDk.

Please advise!
